reactor-core/src/main/java/reactor/core/publisher/MonoDelayUntil.java

```diff
@@ -302,48 +418,161 @@ public Context currentContext() {
 		@Override
 		@Nullable
 		public Object scanUnsafe(Attr key) {
-			if (key == Attr.CANCELLED) return s == Operators.cancelledSubscription();
-			if (key == Attr.PARENT) return s;
-			if (key == Attr.ACTUAL) return parent;
-			if (key == Attr.ERROR) return error;
+			if (key == Attr.CANCELLED) return isTerminated(this.parent.state) && !this.done;
+			if (key == Attr.PARENT) return this.s;
+			if (key == Attr.ACTUAL) return this.parent;
+			if (key == Attr.ERROR) return this.error;
 			if (key == Attr.PREFETCH) return Integer.MAX_VALUE;
 
 			return null;
 		}
 
 		@Override
 		public void onSubscribe(Subscription s) {
-			if (Operators.setOnce(S, this, s)) {
+			if (Operators.validate(this.s, s)) {
+				this.s = s;
+
+				int previousState = markInnerActive();
+				if (isTerminated(previousState)) {
+					s.cancel();
+
+					final DelayUntilCoordinator<?> parent = this.parent;
+					Operators.onDiscard(parent.value, parent.currentContext());
+
+					return;
+				}
+
 				s.request(Long.MAX_VALUE);
-			} else {
-				s.cancel();
 			}
 		}
 
 		@Override
-		public void onNext(Object t) {
+		public void onNext(T t) {
 			//NO-OP
+			Operators.onDiscard(t, this.parent.currentContext());
 		}
 
 		@Override
+		@SuppressWarnings("ConstantConditions")
```

Comment:
it is because of `Exceptions.addThrowable` and `Exceptions.terminate` which can return null. However, we have a strong relationship that Exception.addThrowable happens before any markTermianted in all possible scenarios thus, we have strong guarantee that termination in this case will always return not null value since we called addThrowable before to add the value

Reply:
got it. If possible, it would be preferable to put the annotation on the particular line(s) of code that trigger the warning, even if it means extracting a variable just for that purpose. It is always preferable to a wider suppression on the method, since said widening opens the possibility of hiding legit ConstantConditions in the future. Can you try to do that?

Reply:
done